### ISSUE TITLE
Commander Keen 4,5,6 scrolling

### ISSUE THREAD

**Issue by jschwartzenberg** [stsp/dosemu2#501]
As noted in #491 scrolling in Commander Keen 4,5,6 is not how it should be. It seems I reported this before: https://sourceforge.net/p/dosemu/bugs/481/

There's also the odd KVM issue in #194.

**Comment by jschwartzenberg** [comment]
Crystal Caves seems to have the same issue with the bottom panel shaking when walking.

**Comment by jschwartzenberg** [comment]
The reverse engineering effort of which the sources are on GitHub shows the same issue. It also shows another issue in DOSEMU, CPU usage is very high. (In DOSBox CPU usage is not a problem and the scrolling is fine too.)

The sources can be found here: https://github.com/sulix/omnispeak
To compile for DOS with DJGPP use `make PLATFORM=dos`.

**Comment by stsp** [comment]
Wow, interesting!
Can you reproduce #194 with that?

**Comment by jschwartzenberg** [comment]
No I didn't manage. I guess the difference is that this version runs in protected mode.

@sulix is there a revision of Omnispeak that does not use protected mode in DOS?

I'll have to see if I can reproduce that issue with Keen Dreams. Its original source is available and can be compiled with a Borland toolchain.

**Comment by sulix** [comment]
Unfortunately, there's no real mode version of omnispeak: it's protected mode only. Omnispeak's DOS port is also pretty buggy and incomplete, so I'd be hesitant to use it as an example of how this should be done.

If you're preapred to use Borland C++, the Keen Dreams source should work, as should [CKSRCMOD](https://keenmodding.org/viewtopic.php?t=1926).

I'd hazard a guess as to what the jerky scrolling might be: Keen requires the EGA "HORIZONTAL PEL PANNING" register be properly implemented. A quick grep through the dosemu2 source just shows a (largely unused) [ATTR_HOR_PAN](https://github.com/stsp/dosemu2/search?q=ATTR_HOR_PAN&unscoped_q=ATTR_HOR_PAN) define, which hints to me that it's not supported. Keen uses this as the basis for its smooth horizontal scrolling as can be seen in the [omnispeak code](https://github.com/sulix/omnispeak/blob/master/src/id_vl_dos.c#L697).

**Comment by jschwartzenberg** [comment]
Thanks a lot for the response! I think the scrolling issue is finally clear now. More ID/Apogee games suffer from it and it's great there is source available now.

Regarding the KVM issue, I suspect that the exact compiler used does not matter, as long as the code is running similarly in real mode instead of protected mode. I hope that the Keen Dreams sources will help solving issues with KVM that I ran into. Maybe the Prince of Persia 2 issue (#290) is related as well.

**Comment by stuaxo** [comment]
I seem to recall a nice text mode smooth scrolling demo with PC intern, will try and find it from one of the images on the CD from archive.org

**Comment by stuaxo** [comment]
The version on archive is a bin/cue not easy to mount so posting here for now.

[PCINTERN-EGAVGA-SCROLLING.zip](https://github.com/stsp/dosemu2/files/3505805/PCINTERN-EGAVGA-SCROLLING.zip)

TBH I might have been getting this mixed up with some other text mode demos.

While poking around the net, I found this directory of utilities from SWAG
https://swag.outpostbbs.net/EGAVGA/index.html

Quite a few do VGA scrolling or weird/custom modes that could be worth testing.

**Comment by stsp** [comment]
> Quite a few do VGA scrolling or weird/custom modes that could be worth testing.

Have you tested?

**Comment by stuaxo** [comment]
I'll have a go.

In the meantime I've run the example above, it's a bit quick, so I can't actually tell if it is smooth scrolling (certainly anywhere I pause it it isn't).

**Comment by jschwartzenberg** [comment]
The vertical scrolling is indeed fine, it's just the horizontal scrolling that has the issue. I don't think the KVM bug can be reproduced in Keen Dreams as it misses both a demo mode and the 'Fix jerky motion' option. I hadn't realized that before I was trying.

Hopefully CKSRCMOD will help to reproduce this!

**Comment by stsp** [comment]
Of course everyone here realizes that the
smooth scrolling is not the essential functionality
to bother about right now. We strive to make
the first beta release. The first beta release
doesn't necessarily need the smooth scrolling
and all that - it should barely work. :)
This doesn't mean you shouldn't reproduce
the KVM bug - the more info you collect, the
higher the chances of this to be _eventually_
solved (but not now).

**Comment by jschwartzenberg** [comment]
Yes, I imagine the scrolling is easier to fix now that the cause seems clear.

But the KVM issues can pop up any time, so I hope to be able to reproduce this with something of which source code is available. It's unfortunate the Goodbye Galaxy sources were not released as its newer versions contain the 'Fix jerky motion' option that is needed to reproduce that issue.

**Comment by jschwartzenberg** [comment]
Dr. Who seems to have the same issue.
